pkg/rbd/controllerserver.go

```diff
@@ -368,100 +308,92 @@ func (cs *ControllerServer) ValidateVolumeCapabilities(ctx context.Context, req
+	rbdVol := new(rbdVolume)
+	err := genVolFromVolID(rbdVol, req.GetSourceVolumeId(), req.GetSecrets())
+	if err != nil {
+		if _, ok := err.(ErrImageNotFound); ok {
+			return nil, status.Errorf(codes.NotFound, "Source Volume ID %s not found", req.GetSourceVolumeId())
 		}
-		return nil, status.Errorf(codes.AlreadyExists, "Snapshot with the same name: %s but with different source volume id already exist", req.GetName())
+		return nil, status.Errorf(codes.Internal, err.Error())
 	}
 
-	rbdSnap, err := getRBDSnapshotOptions(req.GetParameters())
-	if err != nil {
-		return nil, status.Error(codes.InvalidArgument, err.Error())
+	// Check if source volume was created with required image features for snaps
+	if !hasSnapshotFeature(rbdVol.ImageFeatures) {
+		return nil, status.Errorf(codes.InvalidArgument, "volume(%s) has not snapshot feature(layering)", req.GetSourceVolumeId())
 	}
 
-	// Generating Snapshot Name and Snapshot ID, as according to CSI spec they MUST be different
-	snapName := req.GetName()
-	uniqueID := uuid.NewUUID().String()
-	rbdVolume, err := getRBDVolumeByID(req.GetSourceVolumeId())
+	// Create snap volume
+	rbdSnap := genSnapFromOptions(rbdVol, req.GetParameters())
+	rbdSnap.RbdImageName = rbdVol.RbdImageName
+	rbdSnap.SizeBytes = rbdVol.VolSize
+	rbdSnap.SourceVolumeID = req.GetSourceVolumeId()
+	rbdSnap.RequestName = req.GetName()
+
+	// Need to check for already existing snapshot name, and if found
+	// check for the requested source volume id and already allocated source volume id
+	found, err := checkSnapExists(rbdSnap, req.GetSecrets())
 	if err != nil {
-		return nil, status.Errorf(codes.NotFound, "Source Volume ID %s cannot found", req.GetSourceVolumeId())
+		if _, ok := err.(ErrSnapNameConflict); ok {
+			return nil, status.Error(codes.AlreadyExists, err.Error())
+		}
+
+		return nil, status.Errorf(codes.Internal, err.Error())
 	}
-	if !hasSnapshotFeature(rbdVolume.ImageFeatures) {
-		return nil, status.Errorf(codes.InvalidArgument, "volume(%s) has not snapshot feature(layering)", req.GetSourceVolumeId())
+	if found {
+		return &csi.CreateSnapshotResponse{
+			Snapshot: &csi.Snapshot{
+				SizeBytes:      rbdSnap.SizeBytes,
+				SnapshotId:     rbdSnap.SnapID,
+				SourceVolumeId: rbdSnap.SourceVolumeID,
+				CreationTime:   rbdSnap.CreatedAt,
+				ReadyToUse:     true,
+			},
+		}, nil
 	}
 
-	rbdSnap.VolName = rbdVolume.VolName
-	rbdSnap.SnapName = snapName
-	snapshotID := "csi-rbd-" + rbdVolume.VolName + "-snap-" + uniqueID
-	rbdSnap.SnapID = snapshotID
-	rbdSnap.SourceVolumeID = req.GetSourceVolumeId()
-	rbdSnap.SizeBytes = rbdVolume.VolSize
-
-	err = cs.doSnapshot(rbdSnap, req.GetSecrets())
-	// if we already have the snapshot, return the snapshot
+	err = reserveSnap(rbdSnap, req.GetSecrets())
 	if err != nil {
 		return nil, status.Error(codes.Internal, err.Error())
 	}
+	defer func() {
+		if err != nil {
+			klog.Warningf("snapshot failed, undoing reservation of snap name: %s", req.GetName())
```

Comment:
Why warning ?

Reply:
Would you prefer Info instead? This path should be arrived at post reservation in case of errors, the original errors are not logged here, but a undo of the reservation is being logged.

Reply:
Removed, not a pattern elsewhere to log, so not adding it here.